cluster/gce/windows/common.psm1

```diff
@@ -146,6 +146,307 @@ function MustDownload-File {
+);
+'@
+$Kernel32 = Add-Type -MemberDefinition $SyscallDefinitions -Name 'Kernel32' -Namespace 'Win32' -PassThru
+
+# ideally we can just use setFilePointer and setEndOfFile to truncate
+
+function Close-Handle {
+  param (
+    [parameter(Mandatory=$true)] [System.IntPtr]$Handle
+  )
+  $ret = $Kernel32::CloseHandle($Handle)
+  $err = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()
+  if (-not $ret) {
+    throw "Failed to close open file handle ${Handle}, system error code: ${err}"
+  }
+}
+
+function Open-File {
+  param (
+    [parameter(Mandatory=$true)] [string]$Path
+  )
+
+  $lpFileName = $Path
+  $dwDesiredAccess = [System.IO.FileAccess]::ReadWrite
+  $dwShareMode = [System.IO.FileShare]::ReadWrite # Fortunately golang also passes these same flags when it creates the log files, so we can open it concurrently.
+  $lpSecurityAttributes = [System.IntPtr]::Zero
+  $dwCreationDisposition = [System.IO.FileMode]::Open
+  $dwFlagsAndAttributes = [System.IO.FileAttributes]::Normal
+  $hTemplateFile = [System.IntPtr]::Zero
+
+  $handle = $Kernel32::CreateFileW($lpFileName, $dwDesiredAccess, $dwShareMode, $lpSecurityAttributes, $dwCreationDisposition, $dwFlagsAndAttributes, $hTemplateFile)
+  $err = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()
+  if ($handle -eq -1) {
+    if ($handle -ne -1) {
+      Close-Handle -Handle $handle
+    }
+    throw "Failed to open file ${Path}, system error code: ${err}"
+  }
+
+  return $handle
+}
+
+# Truncates the file in-place by opening it, moving the file pointer to the beginning,
+# and setting the end of file to the file pointer's location.
+# The file must have been originally created with FILE_SHARE_WRITE for this to be possible.
+# Fortunately Go creates files with FILE_SHARE_READ|FILE_SHARE_WRITE by for all os.Open calls,
+# so our log writers should be doing the right thing.
+function Truncate-File {
+  param (
+    [parameter(Mandatory=$true)] [string]$Path
+  )
+  $INVALID_SET_FILE_POINTER = 0xffffffff
+  $NO_ERROR = 0
+  $FILE_BEGIN = 0
+
+  $handle = Open-File -Path $Path
+
+  # https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-setfilepointer
+  # Docs: Because INVALID_SET_FILE_POINTER is a valid value for the low-order DWORD of the new file pointer,
+  # you must check both the return value of the function and the error code returned by GetLastError to
+  # determine whether or not an error has occurred. If an error has occurred, the return value of SetFilePointer
+  # is INVALID_SET_FILE_POINTER and GetLastError returns a value other than NO_ERROR.
+  $ret = $Kernel32::SetFilePointer($handle, 0, [System.IntPtr]::Zero, $FILE_BEGIN)
+  $err = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()
+  if ($ret -eq $INVALID_SET_FILE_POINTER -and $err -ne $NO_ERROR) {
+    Close-Handle -Handle $handle
+    throw "Failed to set file pointer for handle ${handle}, system error code: ${err}"
+  }
+
+  $ret = $Kernel32::SetEndOfFile($handle)
```

Comment:
Any idea what can happen if the kubelet writes to the file between the SetFilePointer and the SetEndOfFile?

Would SetEndOfFile fail? If so, should we retry with whatever the current FilePointer value is? Or maybe just not throw the exception if SetEndOfFile fails?

Reply:
It should be fine, the file pointer associated with the script's handle is independent from the Kubelet's.